pallets/omnipool-liquidity-mining/src/lib.rs

```diff
@@ -0,0 +1,951 @@
+				lp_position.asset_id,
+			)?;
+
+			if let Some((reward_currency, claimed, _)) = claim_data {
+				if !claimed.is_zero() {
+					Self::deposit_event(Event::RewardClaimed {
+						global_farm_id,
+						yield_farm_id,
+						who: owner.clone(),
+						claimed,
+						reward_currency,
+						deposit_id,
+					});
+				}
+			}
+
+			Self::deposit_event(Event::SharesWithdrawn {
+				global_farm_id,
+				yield_farm_id,
+				who: owner.clone(),
+				amount: withdrawn_amount,
+				deposit_id,
+			});
+
+			if is_destroyed {
+				Self::unlock_lp_postion(deposit_id, &owner)?;
+				<T as pallet::Config>::NFTHandler::burn(
+					&<T as pallet::Config>::NFTCollectionId::get(),
+					&deposit_id,
+					Some(&owner),
+				)?;
+
+				Self::deposit_event(Event::DepositDestroyed { who: owner, deposit_id });
+			}
+
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Account ID of the pot holding all the locked omnipool's positions(NFTs). This account
+	/// is also owner of the NFT collection used to mint liqudity mining's NFTs.
+	fn account_id() -> T::AccountId {
+		<T as pallet::Config>::PalletId::get().into_account_truncating()
+	}
+
+	/// This function transfers omnipool's position NFT to liquidity mining's account. This
+	/// function also saves mapping of the deposit's id to omnipool position's id.
+	fn lock_lp_position(position_id: T::PositionItemId, deposit_id: DepositId) -> Result<(), DispatchError> {
+		<T as pallet::Config>::NFTHandler::transfer(
+			&<T as pallet_omnipool::Config>::NFTCollectionId::get(),
+			&position_id,
+			&Self::account_id(),
+		)?;
+
+		//Mapping of the `deposit_id` to `position_id` so we know which position to unlock when deposit
+		//is destroyed.
+		OmniPositionId::<T>::insert(deposit_id, position_id);
+
+		Ok(())
+	}
+
+	/// This function transfers omnipool's NFT associated with `deposit_id` to `who` and removes
+	/// deposit's id to omnipool position's id for storage.
+	fn unlock_lp_postion(deposit_id: DepositId, who: &T::AccountId) -> Result<(), DispatchError> {
+		OmniPositionId::<T>::try_mutate_exists(deposit_id, |maybe_position_id| -> DispatchResult {
+			//NOTE: not tested, this should never fail
+			let lp_position_id = maybe_position_id
+				.as_mut()
+				.defensive_ok_or::<Error<T>>(InconsistentStateError::MissingLpPosition.into())?;
+
+			<T as pallet::Config>::NFTHandler::transfer(
+				&<T as pallet_omnipool::Config>::NFTCollectionId::get(),
+				lp_position_id,
+				who,
+			)?;
+
+			//NOTE: storage clean up
+			*maybe_position_id = None;
+
+			Ok(())
+		})
+	}
+
+	/// This function returns value of a omnipool's postion in [`LRNA`].
+	fn get_position_value_in_hub_asset(
+		lp_position: &OmniPosition<Balance, T::AssetId>,
+	) -> Result<Balance, DispatchError> {
+		let state = OmnipoolPallet::<T>::load_asset_state(lp_position.asset_id)?;
+
+		state
+			.price()
+			.ok_or(ArithmeticError::DivisionByZero)?
+			.checked_mul_int(lp_position.amount)
+			.ok_or_else(|| ArithmeticError::Overflow.into())
```

Comment:
actually i think, the more precise way would be to use the reserve amounts directly.

instead of using price, you could do amount * hub_reserve and then divide by asset reserve. 
something to property test maybe?